sentry-ruby/lib/sentry/vernier/output.rb

```diff
@@ -0,0 +1,97 @@
+# frozen_string_literal: true
+
+require "json"
+require "rbconfig"
+
+module Sentry
+  module Vernier
+    class Output
+      include Profiler::Helpers
+
+      attr_reader :profile
+
+      def initialize(profile, project_root:, in_app_pattern:, app_dirs_pattern:)
+        @profile = profile
+        @project_root = project_root
+        @in_app_pattern = in_app_pattern
+        @app_dirs_pattern = app_dirs_pattern
+      end
+
+      def to_h
+        @to_h ||= {
+          frames: frames,
+          stacks: stacks,
+          samples: samples,
+          thread_metadata: thread_metadata
+        }
+      end
+
+      private
+
+      def thread_metadata
+        profile.threads.map { |thread_id, thread_info|
+          [thread_id, { name: thread_info[:name] }]
+        }.to_h
+      end
+
+      def samples
+        profile.threads.flat_map { |thread_id, thread_info|
+          started_at = thread_info[:started_at]
+          samples, timestamps = thread_info.values_at(:samples, :timestamps)
+
+          samples.zip(timestamps).map { |stack_id, timestamp|
+            elapsed_since_start_ns = timestamp - started_at
+
+            next if elapsed_since_start_ns < 0
+
+            {
+              thread_id: thread_id.to_s,
+              stack_id: stack_id,
+              elapsed_since_start_ns: elapsed_since_start_ns.to_s
+            }
+          }.compact
+        }
+      end
+
+      def frames
+        funcs = stack_table_hash[:frame_table].fetch(:func)
+        lines = stack_table_hash[:func_table].fetch(:first_line)
+
+        funcs.map { |idx|
+          function, mod = split_module(stack_table_hash[:func_table][:name][idx])
+
+          abs_path = stack_table_hash[:func_table][:filename][idx]
+          in_app = in_app?(abs_path)
+          filename = compute_filename(abs_path, in_app)
+
+          {
+            function: function,
+            module: mod,
+            filename: filename,
+            abs_path: abs_path,
+            lineno: (lineno = lines[idx]) > 0 ? lineno : nil,
+            in_app: in_app
+          }.compact
+        }
+      end
+
+      def stacks
+        sample_stacks = profile.threads.flat_map { |_, thread_info|
+          samples = thread_info[:samples].map { |stack_id| profile.stack(stack_id) }
+
+          samples.map { |stack| [stack.idx, stack.frames.map(&:idx)] }
+        }.to_h
+
+        stacks = profile._stack_table.stack_count.times.map do |stack_id|
```

Comment:
I don't understand this step, are there some stacks that are not in `profile.threads` but are in `profile._stack_table`? Maybe add a comment clarifying the structure.

Reply:
@sl0thentr0py that's what I'm seeing yes:

```
/usr/local/rvm/gems/default/bin/rspec:80
/usr/local/rvm/gems/default/bin/rspec:83
thread-bar-0:88
thread-bar-0:87
thread-bar-0:88
thread-bar-0:87
thread-bar-0:89
thread-bar-0:88
thread-bar-0:87
thread-bar-0:88
thread-bar-0:90
thread-bar-0:87
thread-bar-0:91
thread-bar-0:88
thread-bar-0:87
thread-bar-0:88
thread-bar-0:90
thread-bar-0:87
thread-bar-0:88
thread-bar-0:91
thread-bar-0:87
thread-bar-0:88
thread-bar-0:89
thread-bar-0:88
thread-bar-0:87
thread-bar-0:88
thread-bar-0:90
thread-bar-0:88
thread-bar-0:89
thread-bar-0:88
thread-bar-0:87
thread-bar-0:88
thread-bar-0:87
thread-bar-0:88
thread-bar-0:89
thread-bar-0:88
thread-bar-0:87
thread-bar-0:88
thread-bar-0:93
thread-bar-1:80
thread-bar-1:89
thread-bar-1:90
thread-bar-1:91
thread-bar-1:88
thread-bar-1:87
thread-bar-1:88
thread-bar-1:87
thread-bar-1:91
thread-bar-1:88
thread-bar-1:89
thread-bar-1:88
thread-bar-1:89
thread-bar-1:88
thread-bar-1:89
thread-bar-1:87
thread-bar-1:88
thread-bar-1:91
thread-bar-1:88
thread-bar-1:87
thread-bar-1:88
thread-bar-1:90
thread-bar-1:87
thread-bar-1:90
thread-bar-1:88
thread-bar-1:90
thread-bar-1:88
thread-bar-1:91
thread-bar-1:87
thread-bar-1:89
thread-bar-1:88
thread-bar-1:89
thread-bar-1:91
thread-bar-1:88
thread-bar-1:87
thread-bar-1:88
thread-bar-1:91
thread-bar-1:87
thread-bar-1:93
Did not find stack 0 in sample stacks
Did not find stack 1 in sample stacks
Did not find stack 2 in sample stacks
Did not find stack 3 in sample stacks
Did not find stack 4 in sample stacks
Did not find stack 5 in sample stacks
Did not find stack 6 in sample stacks
Did not find stack 7 in sample stacks
Did not find stack 8 in sample stacks
Did not find stack 9 in sample stacks
Did not find stack 10 in sample stacks
Did not find stack 11 in sample stacks
Did not find stack 12 in sample stacks
Did not find stack 13 in sample stacks
Did not find stack 14 in sample stacks
Did not find stack 15 in sample stacks
Did not find stack 16 in sample stacks
Did not find stack 17 in sample stacks
Did not find stack 18 in sample stacks
Did not find stack 19 in sample stacks
Did not find stack 20 in sample stacks
Did not find stack 21 in sample stacks
Did not find stack 22 in sample stacks
Did not find stack 23 in sample stacks
Did not find stack 24 in sample stacks
Did not find stack 25 in sample stacks
Did not find stack 26 in sample stacks
Did not find stack 27 in sample stacks
Did not find stack 28 in sample stacks
Did not find stack 29 in sample stacks
Did not find stack 30 in sample stacks
Did not find stack 31 in sample stacks
Did not find stack 32 in sample stacks
Did not find stack 33 in sample stacks
Did not find stack 34 in sample stacks
Did not find stack 35 in sample stacks
Did not find stack 36 in sample stacks
Did not find stack 37 in sample stacks
Did not find stack 38 in sample stacks
Did not find stack 39 in sample stacks
Did not find stack 40 in sample stacks
Did not find stack 41 in sample stacks
Did not find stack 42 in sample stacks
Did not find stack 43 in sample stacks
Did not find stack 44 in sample stacks
Did not find stack 45 in sample stacks
Did not find stack 46 in sample stacks
Did not find stack 47 in sample stacks
Did not find stack 48 in sample stacks
Did not find stack 49 in sample stacks
Did not find stack 50 in sample stacks
Did not find stack 51 in sample stacks
Did not find stack 52 in sample stacks
Did not find stack 53 in sample stacks
Did not find stack 54 in sample stacks
Did not find stack 55 in sample stacks
Did not find stack 56 in sample stacks
Did not find stack 57 in sample stacks
Did not find stack 58 in sample stacks
Did not find stack 59 in sample stacks
Did not find stack 60 in sample stacks
Did not find stack 61 in sample stacks
Did not find stack 62 in sample stacks
Did not find stack 63 in sample stacks
Did not find stack 64 in sample stacks
Did not find stack 65 in sample stacks
Did not find stack 66 in sample stacks
Did not find stack 67 in sample stacks
Did not find stack 68 in sample stacks
Did not find stack 69 in sample stacks
Did not find stack 70 in sample stacks
Did not find stack 71 in sample stacks
Did not find stack 72 in sample stacks
Did not find stack 73 in sample stacks
Did not find stack 74 in sample stacks
Did not find stack 75 in sample stacks
Did not find stack 76 in sample stacks
Did not find stack 77 in sample stacks
Did not find stack 78 in sample stacks
Did not find stack 79 in sample stacks
Found stack 80 in sample stacks
Did not find stack 81 in sample stacks
Did not find stack 82 in sample stacks
Found stack 83 in sample stacks
Did not find stack 84 in sample stacks
Did not find stack 85 in sample stacks
Did not find stack 86 in sample stacks
Found stack 87 in sample stacks
Found stack 88 in sample stacks
Found stack 89 in sample stacks
Found stack 90 in sample stacks
Found stack 91 in sample stacks
Did not find stack 92 in sample stacks
Found stack 93 in sample stacks
```